src/gt4py/next/iterator/embedded.py

```diff
@@ -1114,6 +1102,108 @@ def impl(it: ItIterator) -> ItIterator:
     return impl
 
 
+DT = TypeVar("DT")
+
+
+class _List(tuple, Generic[DT]):  # TODO encode size in type?
+    ...
+
+
+@dataclasses.dataclass(frozen=True)
+class _ConstList(Generic[DT]):
+    value: DT
+
+    def __getitem__(self, _):
+        return self.value
+
+
+@builtins.neighbors.register(EMBEDDED)
+def neighbors(offset: runtime.Offset, it: ItIterator) -> _List:
+    # TODO offset provider should be part of the iterator interface?
+    offset_str = offset.value if isinstance(offset, runtime.Offset) else offset
+    return _List(
+        builtins.deref(it.shift(offset_str, i))
+        if builtins.can_deref(it.shift(offset_str, i))
+        else None
+        for i in range(it.offset_provider[offset_str].max_neighbors)  # type: ignore
+    )
+
+
+@builtins.list_get.register(EMBEDDED)
+def list_get(i, lst: _List[Optional[DT]]) -> Optional[DT]:
+    return lst[i]
+
+
+@builtins.map_.register(EMBEDDED)
+def map_(op):
+    def impl_(*lists):
+        return _List(map(lambda x: op(*x), zip(*lists)))
+
+    return impl_
+
+
+@builtins.make_const_list.register(EMBEDDED)
+def make_const_list(value):
+    return _ConstList(value)
+
+
+@builtins.reduce.register(EMBEDDED)
+def reduce(fun, init):
+    def sten(*lists):
+        # TODO: assert check_that_all_iterators_are_compatible(*iters)
+        lst = None
+        for cur in lists:
+            if isinstance(cur, _List):
+                lst = cur
+                break
+        n = len(lst)
+        res = init
+        for i in range(n):
+            # we can check a single argument
+            # because all arguments share the same pattern
+            if lst[i] is None:
+                break
+            res = fun(
+                res,
+                *(lst[i] for lst in lists),
+            )
+        return res
+
+    return sten
+
+
+@dataclasses.dataclass(frozen=True)
+class SparseListIterator:
+    it: ItIterator
+    list_offset: Tag
+    offsets: Sequence[OffsetPart] = dataclasses.field(default_factory=list, kw_only=True)
+
+    @property
+    def offset_provider(self):
+        return self.it.offset_provider
+
+    def deref(self) -> Any:
+        return _List(
+            builtins.deref(builtins.shift(*self.offsets, SparseTag(self.list_offset), i)(self.it))
+            if builtins.can_deref(
+                builtins.shift(*self.offsets, SparseTag(self.list_offset), i)(self.it)
+            )
+            else None
+            for i in range(self.it.offset_provider[self.list_offset].max_neighbors)  # type: ignore
```

Comment:
Same rationale of previous comments by @fthaler of using lists of different lengths instead of using lists with optionals:
```suggestion
            builtins.deref(shifted_it)
            for i in range(self.it.offset_provider[self.list_offset].max_neighbors)  # type: ignore
            if builtins.can_deref(
                shifted_it := (builtins.shift(*self.offsets, SparseTag(self.list_offset), i)(self.it))
            )
```

Reply:
done